src/lib.rs

```diff
@@ -45,115 +46,300 @@ pub enum Protection {
+/// let file = File::open("README.md").unwrap();
+/// let mmap = unsafe { memmap::file(&file).map().unwrap() };
+/// assert_eq!(b"# memmap", &mmap[0..8]);
 /// ```
+///
+/// See [`MmapMut`](struct.MmapMut.html) for the mutable version.
 pub struct Mmap {
     inner: MmapInner
 }
 
 impl Mmap {
-
-    /// Opens a file-backed memory map.
+    /// Change the `Protection` this mapping was created with.
     ///
-    /// The file must be opened with read permissions, and write permissions if
-    /// the supplied protection is `ReadWrite`. The file must not be empty.
-    pub fn open(file: &File, prot: Protection) -> Result<Mmap> {
-        let len = try!(file.metadata()).len();
-        if len > usize::MAX as u64 {
-            return Err(Error::new(ErrorKind::InvalidData,
-                                  "file length overflows usize"));
-        }
-        MmapInner::open(file, prot, 0, len as usize).map(|inner| Mmap { inner: inner })
+    /// This method only changes the protection of the underlying mapping,
+    /// but it doesn't make an `MmapMut` from an `Mmap`, use [`make_mut()`](#method.make_mut)
+    /// method for that.
+    ///
+    /// If you create a read-only file-backed mapping, you can **not** use this method to make the
+    /// mapping writeable. Remap the file instead.
+    ///
+    /// # Errors
+    ///
+    /// This method returns `Err` when the underlying system call fails, which can happen for
+    /// a variety of reasons, such as when you don't have the necessary permissions for the file.
+    pub fn set_protection(&mut self, protection: Protection) -> Result<()> {
+        self.inner.set_protection(protection)
     }
 
-    /// Opens a file-backed memory map.
+    /// Change the `Protection` this mapping was created with to make it mutable.
+    ///
+    /// If you create a read-only file-backed mapping, you can **not** use this method to make the
+    /// mapping writeable. Remap the file instead.
+    ///
+    /// # Errors
+    ///
+    /// This method returns `Err` when the underlying system call fails, which can happen for
+    /// a variety of reasons, such as when you don't have the necessary permissions for the file.
     ///
-    /// The file must not be empty.
-    pub fn open_path<P>(path: P, prot: Protection) -> Result<Mmap>
-    where P: AsRef<Path> {
-        let file = try!(prot.as_open_options().open(path));
-        let len = try!(file.metadata()).len();
-        if len > usize::MAX as u64 {
-            return Err(Error::new(ErrorKind::InvalidData,
-                                  "file length overflows usize"));
+    /// This method *also* returns `Err` with `ErrorKind` set to `InvalidInput` if the specified
+    /// protection does not allow the mapping to be mutable.
+    pub fn make_mut(mut self, protection: Protection) -> Result<MmapMut> {
+        try!(self.inner.set_protection(protection));
+        match protection {
+            Protection::Read | Protection::ReadExecute => Err(Error::new(
```

Comment:
IMO at this point it should just try to `set_protection` the mapping to the +w and only return Err if *that* fails.